### ISSUE TITLE
Auto API Controllers for dotnetcore

### ISSUE THREAD

**Issue by carlospence** [MrCMS/MrCMS#102]
Hello, Good evening

I am thinking of a feature in which Web API's can be automatically generated from Services (Auto API Controllers) and be display using Swagger for standalone front end consumption (SPA or Mobile)

I am looking @making MrCMS create APIs with ease either directly from APIController or from Services and they can be displayed using Swagger either at the backend (with authentication) or front-end (without authentication)

Just want to know if @gmstr or @Will-Bill  have any such feature in the works

**Comment by Will-Bill** [comment]
This sounds great, keep me posted. It's not something we are working on at the moment.

In the DNC version, one thing I am not happy with is the embedded views/content within apps. It requires the app to be rebuilt either manually or using dotnet watch, but neither are nice as just running the app and refreshing browser and using runtime compilation. I think we need to do some work to try a different method of registering apps so razor is not so clunky.

Having and API would enable us to have any front end, and that in itself is a great idea.

**Comment by carlospence** [comment]
Yes, I thought of the embedded views and contents too, they are too clunky and will make the final dll files to be heavy. Is there a way to look at Razor Pages instead of pure MVC for the apps. They have one thing "Single Responsibility" and if we look further, we might be able to avoid the rebuilding of the apps (though not too sure about this) but it is something worth looking into.

I constantly use dotnet watch. 

I think we should think of a way to register apps like a plugin or run the apps like a microservice. 

I sometimes don't want to use asp.net for my frontend or want to delegate that to another team using another frontend platform like angular or react or vue, with the API, that can easily be consumed. I added JWT to the previous version that I used and will also add to this.

I have just gotten Swagger to work.

**Comment by carlospence** [comment]
Thinking about registration of apps. I think there should be two ways of registering apps 

1) This new DNC version for development and testing
2) Previous MVC 5 version with some advancement for real-time deployment

**Previous MVC 5 version with some advancement**

This time around there would be a apps folder or plugin folder inside the MrCMS.web that will contain all the plugins. 
There would be some specifications on how each unique plugin folders will be layered or configured.
Each plugin could be a zipped package that will be extracted automatically by the CMS.

For example, the apps I have built using MrCMS MVC5 all have their libraries in a separate project.
The app in the Apps folder only contains the Controllers, Views, Contents and scripts. 

All my entities, services. Dtos, Events Models etc  are in a separate library file and MrCMS automatically  see them by me just add MrCMSAssemblyAttribute to the library AssemblyInfo
e.g [assembly: MrCMSAssembly(SyncBusLibInfo.AssemblyFileVersion)]

So, am thinking of us using this method that is already possible in the previous version and advance it to do auto addition of plugins without rebuilding. The Apps or plugin will just meet some necessary condition exposed by MrCMS plugin API. Though it is a tough call I think it is doable

**Comment by kschost100** [comment]
> In the DNC version, one thing I am not happy with is the embedded views/content within apps. It requires the app to be rebuilt either manually or using dotnet watch, but neither are nice as just running the app and refreshing browser and using runtime compilation. I think we need to do some work to try a different method of registering apps so Razor is not so clunky.

@Will-Bill I can work on the building and publish. Would you like to move the Admin views back to the Apps.Admin project, as I can have all the App projects under an Apps folder without the views being embedded. We can also do the same for the Themes. Let me know.

**Comment by Will-Bill** [comment]
@kschost100 sure can do, what is the method?

**Comment by kschost100** [comment]
@Will-Bill 
Because the Apps are referenced within the web project DNC will always build them to the default folder. What we do is build App projects to App folder within the web project and then delete after publishing. this will allow us to build the Apps with Views, and WWWRoot as unembedded objects. I do have this working with Core and Theme as I am creating a theme for the project, It did take me some time to research and from what I see DNC 2.0 doesn't play very nice with plugins, but I see 3.0 is coming with support for it. I can do more research towards a better solution, as I know the publish and delete is a hack but it is working for me. I will do this for all the apps and create a script to perform the delete function. how would you like to get this, New branch maybe so you can test it out.

**Comment by carlospence** [comment]
@Will Bill

In reaching the goal of Auto API Controllers from services, I have reached the following milestones.

1) Integration of Swagger into MrCMS

![image](https://user-images.githubusercontent.com/12647064/65388871-d48df380-dd47-11e9-8d55-e15669dbea89.png)


2) Integration of IdentityServer4 into MrCMS

![image](https://user-images.githubusercontent.com/12647064/65388883-ef606800-dd47-11e9-9e84-4cc5621e2896.png)


3) Adding a MrCMS App (more of a library) for a dedicated MrCMS ApiControllers

In achieving No. 1
Some Controller Actions in MrCMS needed to be given the right Httpverb of either post, get or put.

In achieving No. 2
I ported the following open-sourced libraries into MrCMS

albertodall/IdentityServer4.NHibernate- (https://github.com/albertodall/IdentityServer4.NHibernate)
skoruba/IdentityServer4.Admin- (https://github.com/skoruba/IdentityServer4.Admin)

There are 2 quirk things I encountered while developing the apps above.

i) Registration of all the apps in the main MrCMS.web, even if the views are embedded as it is, I think auto-registration of apps should be available i.e MrCMS will auto-register the apps but the Administrator will enable the usage of the app. With or without auto-registration, apps can be packaged as a nuget package

ii) I had to register my TagHelpers in MrCMS.web.admin. it meant even if an auto-registration of Apps is achieved, Taghelpers will still have to be manually added to ..\MrCMS.Web\Areas\Admin\Views\_ViewImports.cshtml and even if the files are finally moved into MrCMS.Web.Apps.Admin, there must be a way to also auto-register Taghelpers.

I created a fork @https://github.com/carlospence/MrCMS under a branch called experimental. Would want you to check it out because I had an issue somewhere and I believe you might be able to sort it out. Let me know so that I can create a PR

I would love your input to solve an issue and have your opinion about what I have done so far before I move to the final part of my milestone because I believe I have laid the foundation for the final Auto-WebAPI generation from Services.

**Comment by Will-Bill** [comment]
Thank you for this - we are a little busy on some internal projects but will be picking things up as soon as we can.

**Comment by Will-Bill** [comment]
@carlospence are you on Skype? Could you add me if so to discuss? Username: thought.will
